setup_dev.py

```diff
@@ -0,0 +1,532 @@
+
+  subparsers = parser.add_subparsers(dest="command", title='sub-command')
+  parser_toolchain = subparsers.add_parser("toolchain",
+                                           help="Regenerate toolchain config")
+  args = parser.parse_args()
+  return args
+
+
+def log_verbose(msg: str):
+  print(msg)
+
+
+def die(msg: str):
+  print(msg)
+  sys.exit(1)
+
+
+def find_command(cmd: str) -> Optional[str]:
+  p = shutil.which(cmd)
+  if p:
+    log_verbose(f"- found {cmd}: {p}")
+  else:
+    log_verbose(f"- not found {cmd}")
+  return p
+
+
+def setup_toolchain(args, config: Config):
+  # c/c++ compiler
+  cc_exe = args.cc
+  cxx_exe = args.cxx
+  # Default to clang if omitted entirely.
+  if cc_exe is None:
+    cc_exe = find_command("clang")
+  if cxx_exe is None:
+    cxx_exe = find_command("clang++")
+
+  # ccache
+  ccache_path = None
+  if args.ccache or args.ccache is None:
+    ccache_path = find_command("ccache")
+    if args.ccache and not ccache_path:
+      die("ERROR: ccache enabled but not found. "
+          "Ensure that 'ccache' is on the path")
+
+  # lld
+  enable_lld = False
+  if args.lld or args.lld is None:
+    enable_lld = bool(find_command("lld"))
+    if args.lld and not enable_lld:
+      die("ERROR: lld enabled but not found. Ensure that 'lld' is on your path")
+
+  # dev mode
+  if args.dev:
+    log_verbose("- dev mode enabled (artifacts cannot be relocated)")
+  config.write_toolchain_config(dev_mode=args.dev,
+                                cc_exe=cc_exe,
+                                cxx_exe=cxx_exe,
+                                python_exe=sys.executable,
+                                ccache_exe=ccache_path,
+                                enable_lld=enable_lld)
+
+
+def build_external_llvm(args, config: Config):
+  # TODO: Make these configurable.
+  llvm_main_src_dir = repo_root / "third_party" / "llvm-project"
+  llvm_build_dir = config.llvm_build_dir
+  mlir_build_dir = config.mlir_build_dir
+  llvm_build_dir.mkdir(parents=True, exist_ok=True)
+  mlir_build_dir.mkdir(parents=True, exist_ok=True)
+
+  # Build type
+  build_type_args = build_config_to_args(config.llvm_build_config, "llvm")
+
+  # LLVM configure step.
+  if args.llvm_configure:
+    llvm_configure_cmake_args = [
+        "cmake",
+        "-GNinja",
+        "-S",
+        str(llvm_main_src_dir / "llvm"),
+        "-B",
+        str(llvm_build_dir),
+        f"-DCMAKE_PROJECT_TOP_LEVEL_INCLUDES={config.get_config_path(config.toolchain_config_file_name)}",
+        f"-DBUILD_SHARED_LIBS={'ON' if args.llvm_shared else 'OFF'}",
+        "-DLLVM_ENABLE_PROJECTS=llvm;lld;clang",
+        "-DLLVM_TARGETS_TO_BUILD=X86;ARM;AArch64;RISCV;NVPTX;WebAssembly;AMDGPU",
+        "-DLLVM_BUILD_EXAMPLES=OFF",
+        "-DLLVM_INSTALL_UTILS=ON",
+    ]
+    llvm_configure_cmake_args.extend(build_type_args)
+    log_verbose(f"+ Configuring LLVM: {' '.join(llvm_configure_cmake_args)}")
+    subprocess.check_call(llvm_configure_cmake_args)
```

Comment:
`check_call` and friends are no longer in vogue. It's all about `subprocess.run` these days. I think this is actually the right call: The `check_call` family offered a weird subset of the API that didn't flow gracefully between use cases (https://docs.python.org/3/library/subprocess.html#older-high-level-api).

```suggestion
  subprocess.run(llvm_configure_cmake_args, check=True)
```

Slightly more verbose in this case but way easier to change to e.g. capture the output or get the return coe